src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,344 @@
+        return body;
+    }
+
+    Stmt visit(const Allocate *op) override {
+        // If this Allocate node is allocating a buffer for a producer,
+        // and there is a Store node inside of an Atomic node requiring mutex lock
+        // matching the name of the Allocate, allocate a mutex lock.
+        FindStoreInAtomicMutex finder({op->name});
+        op->body.accept(&finder);
+        if (!finder.found) {
+            // No Atomic node that requires mutex lock from this node inside.
+            return IRMutator::visit(op);
+        }
+
+        if (allocated_mutexes.find(finder.mutex_name) != allocated_mutexes.end()) {
+            // We've already allocated a mutex.
+            return IRMutator::visit(op);
+        }
+
+        allocated_mutexes.insert(finder.mutex_name);
+
+        const std::string &mutex_name = finder.mutex_name;
+        Stmt body = mutate(op->body);
+        Expr extent = Expr(1);
+        for (Expr e : op->extents) {
+            extent = extent * e;
+        }
+        body = allocate_mutex(mutex_name, extent, body);
+        return Allocate::make(op->name,
+                              op->type,
+                              op->memory_type,
+                              op->extents,
+                              op->condition,
+                              std::move(body),
+                              op->new_expr,
+                              op->free_function);
+    }
+
+    Stmt visit(const ProducerConsumer *op) override {
+        // Usually we allocate the mutex buffer at the Allocate node,
+        // but outputs don't have Allocate. For those we allocate the mutex
+        // buffer at the producer node.
+
+        if (!op->is_producer) {
+            // This is a consumer.
+            return IRMutator::visit(op);
+        }
+
+        // Find the corresponding output.
+        auto func_it = env.find(op->name);
+        const Function &f = func_it->second;
+        internal_assert(f.output_buffers().size() > 0) <<
+            "Found a producer node that contains an atomic node that requires mutex lock, "
+            "but does not have an Allocate node and is not an output function. This is not supported.\n";
+
+        std::set<std::string> store_names;
+        for (auto buffer : f.output_buffers()) {
+            store_names.insert(buffer.name());
+        }
+
+        FindStoreInAtomicMutex finder(store_names);
+        op->body.accept(&finder);
+        if (!finder.found) {
+            // No Atomic node that requires mutex lock from this node inside.
+            return IRMutator::visit(op);
+        }
+
+        if (allocated_mutexes.find(finder.mutex_name) != allocated_mutexes.end()) {
+            // We've already allocated a mutex.
+            return IRMutator::visit(op);
+        }
+
+        allocated_mutexes.insert(finder.mutex_name);
+
+        // We assume all output buffers have the same extent
+        Parameter output_buffer = f.output_buffers()[0];
```

Comment:
Change Parameter to OutputImageParam and you'll get your extent methods to use below.

Reply:
Done! Can't believe I miss this (probably because Function doesn't expose it and it's only in Func).